pkg/source/pulsar.go

```diff
@@ -286,9 +286,41 @@ func newAckTracker(size uint) *ackTracker {
 }
 
 func (t *ackTracker) ack(batchID int) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
 	if batchID < 0 {
 		return true
 	}
 	t.batchIDs.Clear(uint(batchID))
 	return t.batchIDs.None()
 }
+
+type ackTrackers struct {
+	trackers sync.Map
+}
+
+func newAckTrackers() *ackTrackers {
+	return &ackTrackers{}
+}
+
+func (a *ackTrackers) key(msg pulsar.MessageID) string {
+	return fmt.Sprintf("%d:%d", msg.LedgerID(), msg.EntryID())
+}
+
+func (a *ackTrackers) tryAdd(msg pulsar.MessageID) bool {
+	_, loaded := a.trackers.LoadOrStore(a.key(msg), newAckTracker(uint(msg.BatchSize())))
```

Comment:
But doing newAckTracker(uint(msg.BatchSize()) every time won’t be too costly?

Reply:
good point